space_charge_class_electromagnetic.py

```diff
@@ -0,0 +1,115 @@
+#     and documentation freely and without charge.
+#
+#     CERN undertakes no obligation  for  the  maintenance  of  this
+#     program,  nor responsibility for its correctness,  and accepts
+#     no liability whatsoever resulting from its use.
+#
+#     Program  and documentation are provided solely for the use  of
+#     the organization to which they are distributed.
+#
+#     This program  may  not  be  copied  or  otherwise  distributed
+#     without  permission. This message must be retained on this and
+#     any other authorized copies.
+#
+#     The material cannot be sold. CERN should be  given  credit  in
+#     all references.
+#
+#-End-preamble---------------------------------------------------------
+
+import numpy as np
+from space_charge_class import space_charge
+from scipy.constants import c, epsilon_0, mu_0
+
+na = lambda x: np.array([x])
+
+
+class space_charge_electromagnetic(space_charge, object):
+
+    def __init__(self, chamb, Dh, Dt_sc=None, PyPICmode='FiniteDifferences_ShortleyWeller' , sparse_solver='scipy_slu',
+                 f_telescope=None, target_grid=None, N_nodes_discard=None, N_min_Dh_main=None, gamma = 479.):
+
+        super(space_charge_electromagnetic, self).__init__(chamb, Dh, Dt_sc, PyPICmode , sparse_solver,
+                     f_telescope, target_grid, N_nodes_discard, N_min_Dh_main)
+
+        self.state_Ax = self.PyPICobj.get_state_object()
+        self.state_Ay = self.PyPICobj.get_state_object()
+
+        self.gamma = gamma
+        self.beta = np.sqrt(1-1/(gamma*gamma))
+
+    def recompute_spchg_efield(self, MP_e, flag_solve=True, flag_reset=True):
+        # scatter rho
+        self.PyPICobj.scatter(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp], MP_e.nel_mp[0:MP_e.N_mp],
+                charge=MP_e.charge, flag_add=not(flag_reset))
+
+        # scatter currents
+        self.state_Ax.scatter(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp],
+                epsilon_0 * mu_0 * MP_e.nel_mp[0:MP_e.N_mp] * MP_e.vx_mp[0:MP_e.N_mp],
+                charge=MP_e.charge, flag_add=not(flag_reset))
+        self.state_Ay.scatter(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp],
+                 epsilon_0 * mu_0 * MP_e.nel_mp[0:MP_e.N_mp] * MP_e.vy_mp[0:MP_e.N_mp],
+                charge=MP_e.charge, flag_add=not(flag_reset))
+
+        # solve
+        if flag_solve:
+            self.PyPICobj.solve()
+            self.PyPICobj.solve_states([self.state_Ax, self.state_Ay])
+
+
+    def get_sc_b_field(self, MP_e):
+
+        _, dAx_dy = self.state_Ax.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+        dAy_dx, _ = self.state_Ay.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+        dphi_dx, dphi_dy = self.PyPICobj.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+
+        Ex_prime = self.gamma*dphi_dx
+        Ey_prime = self.gamma*dphi_dy
+
+        dAz_dx = -self.gamma*self.beta*c*epsilon_0*mu_0*dphi_dx
```

Comment:
@lgiacome Let's try to be consistent with the notation of the note: distinguishing phi from Phi